[HOLD] - Doc fixes (#10)
mahmutcanga/java-banking-pubsub-dapr-sample: README.md

```

To check if Dapr is installed, run the following command:

```bash
kubectl get pods -n dapr-system
```

This will output something like this:

```bash
NAME                                    READY   STATUS    RESTARTS   AGE
dapr-dashboard-7798c78c74-4ddw5         1/1     Running   0          5m9s
dapr-operator-8955455bf-6qs79           1/1     Running   0          5m9s
dapr-placement-server-0                 1/1     Running   0          5m9s
dapr-sentry-69c8464dc9-6rtkf            1/1     Running   0          5m9s
dapr-sidecar-injector-fcbc4b979-wcjmj   1/1     Running   0          5m9s
```

We will deploy the application to its own namespace. Run following command to create a namespace for the application:

```bash
kubectl create namespace $AZURE_ENV_NAME
```

Check if the namespace is created:

```bash
kubectl get namespace
```

This will output something like this:

```bash
NAME                                  STATUS   AGE
dapr-system                           Active   6m48s
default                               Active   122m
gatekeeper-system                     Active   121m

> java-banking-pubsub-dapr-sample-dev   Active   25s

Reviewer: this is based on the `$AZURE_ENV_NAME` we select. Worth putting a comment here explaining that this variable will be listed as a namespace instead of your local example?

Author: yeah, that makes sense.

kube-node-lease                       Active   122m
kube-public                           Active   122m
kube-system                           Active   122m
```

We are using an in-cluster Redis instance for the pub-sub and state store components. To deploy Redis, run the following command:

```bash
kubectl apply -f ./infra/redis.yaml --namespace $AZURE_ENV_NAME --wait=true
```

You can check the Redis installation by running the following command:

```bash
kubectl get pods --namespace $AZURE_ENV_NAME
```

This will output something like this:

```bash
NAME               READY   STATUS    RESTARTS      AGE
redis-master-0     1/1     Running   0             2m50s
redis-replicas-0   0/1     Running   3 (28s ago)   2m50s
```

### 4. Deploy PubSub and State Store components

Now, we have an in-cluster Redis instance running. We can deploy the pub-sub and state store components for Dapr.

#### 4.1. Deploy pub-sub broker component backed by Redis

```bash
kubectl apply -f ./local/components/pubsub.yaml --wait=true --namespace $AZURE_ENV_NAME
```

You can check the pub-sub broker installation by running the following command:

```bash

> dapr components -k --namespace $AZURE_ENV_NAME

Reviewer: is this the correct command?
the one I was using is `kubectl get component --namespace $AZURE_ENV_NAME`

If the outputs are the same, just ignore this comment

Author: assuming you have `dapr` cli installed in your environment 😅

```

This will output something like this:

```bash
NAMESPACE                            NAME                   TYPE          VERSION  SCOPES  CREATED              AGE
java-banking-pubsub-dapr-sample-dev  money-transfer-pubsub  pubsub.redis  v1               2023-05-22 16:55.50  56s
```

#### 4.2. Deploy state store component backed Redis

```bash
kubectl apply -f ./local/components/state.yaml --wait=true --namespace $AZURE_ENV_NAME
```

> You can check the pub-sub broker installation by running the following command:

Reviewer: ```suggestion
You can check the state store component installation by running the following command:
```

```bash
dapr components -k --namespace $AZURE_ENV_NAME
```

This will output something like this:

```bash
NAMESPACE                            NAME                   TYPE          VERSION  SCOPES  CREATED              AGE
java-banking-pubsub-dapr-sample-dev  money-transfer-pubsub  pubsub.redis  v1               2023-05-22 16:55.50  2m
java-banking-pubsub-dapr-sample-dev  money-transfer-state   state.redis   v1               2023-05-22 16:57.46  13s
```

### 5. Deploy microservices

Previously, we have setup our cluster, deployed Dapr and its components. Now, we can deploy the microservices.

Before you start the deployment, you will need to login to Azure Container Registry (ACR) to be able to push the images. To do so, run the following command.
This command logs in to an Azure Container Registry through the Docker CLI. Docker must be installed on your machine. Once done, use 'docker logout <registry url>' to log out. (If you only need an access token and do not want to install Docker, specify '--expose-token').

```bash
az acr login --name $AZURE_CONTAINER_REGISTRY_NAME
```

This will output something like this:

```bash
Login Succeeded
```

> This command deploys all services defined in `azure.yaml` files in the namespace defined by the environment name, based on their individual deployment under `manifests` folder by `azd` convention.

Reviewer: ```suggestion
This command deploys all services defined in `azure.yaml` file in the namespace defined by the environment name, based on their individual deployment under `manifests` folder by `azd` convention.
```

```bash
NAME                 TYPE           CLUSTER-IP   EXTERNAL-IP    PORT(S)        AGE
public-api-service   LoadBalancer   10.0.54.16   20.8.255.153   80:31319/TCP   2m27s
```

You can use the `EXTERNAL-IP` to access the API. You can run the following command to set the API endpoint available as environment variable:

```bash
PUBLIC_API_SERVICE=$(k get svc public-api-service -n $AZURE_ENV_NAME -o=jsonpath='{.status.loadBalancer.ingress[0].ip}')
```

To check if the API is running and healthy, run the following command:

```bash
curl $PUBLIC_API_SERVICE
```

This will output something like this:

```bash
{"message":"Public API Service Started, version: 1.0.0"}
```

### 7. Test the application

First, you need to create a new account for a user:

```curl
curl -X POST \
  $PUBLIC_API_SERVICE/accounts \
  -H 'Content-Type: application/json' \
  -d '{
    "owner": "A",
    "amount": 100
}'
```

An example request to start a new transfer workflow is:

```curl
curl -X POST \
  $PUBLIC_API_SERVICE/transfers \
  -H 'Content-Type: application/json' \
  -d '{
    "sender": "A",
    "receiver": "B",
    "amount": 100
}'
```

You can query the status of a transfer:

```curl
curl -X GET \

>   http://localhost:8080/transfers/{transferId} \

Reviewer: ```suggestion
  $PUBLIC_API_SERVICE/transfers/{transferId} \
```

│      └── config.json         [ Contains environment configuration ]
```

> ### 2. Up & Running

Reviewer: Suggestion: 

```suggestion
### 2. Resource Provision and Deploy
```

namespace/java-banking-pubsub-dapr-sample created

🚀 Deploy Redis on AKS

serviceaccount/redis created
secret/redis created
configmap/redis-configuration created
configmap/redis-health created
configmap/redis-scripts created
service/redis-headless created
service/redis-master created
service/redis-replicas created
statefulset.apps/redis-master created
statefulset.apps/redis-replicas created

🚀 Deploy pub-sub broker component backed by Redis

component.dapr.io/money-transfer-pubsub created

🚀 Deploy state store component backed Redis

component.dapr.io/money-transfer-state created

Deploying services (azd deploy)

  (✓) Done: Deploying service account-service
  - Endpoint: http://None:80, (Service, Type: ClusterIP)

  (✓) Done: Deploying service fraud-service
  - Endpoint: http://None:80, (Service, Type: ClusterIP)

  (✓) Done: Deploying service notification-service
  - Endpoint: http://None:80, (Service, Type: ClusterIP)

  (✓) Done: Deploying service public-api-service
  - Endpoint: http://20.101.13.7, (Service, Type: LoadBalancer)

SUCCESS: Your application was provisioned and deployed to Azure in 16 minutes 19 seconds.

```

> The Azure Developer CLI supports various extension points to customize your workflows and deployments. The hooks middleware allows you to execute custom scripts before and after azd commands and service lifecycle events. In our setup, we are using `postprovision` hook to install:

Reviewer: Nit:
Include that the hook definition|declaration is made in the `azure.yaml` file ?
